nfsv4/xdr.c

```diff
@@ -0,0 +1,174 @@
+    r->b->end += NFS4_SESSIONID_SIZE;
+}
+
+void push_client_id(rpc r)
+{
+    buffer_extend(r->b, sizeof(clientid));
+    memcpy(r->b->contents + r->b->end, &r->c->clientid, sizeof(clientid));
+    r->b->end += sizeof(r->c->clientid);
+}
+
+
+// section 18.36, pp 513, rfc 5661
+#define CREATE_SESSION4_FLAG_PERSIST 1
+void push_create_session(rpc r)
+{
+    push_op(r, OP_CREATE_SESSION);
+    push_client_id(r);
+
+    push_be32(r->b, r->c->sequence);
+    push_be32(r->b, CREATE_SESSION4_FLAG_PERSIST);
+    push_channel_attrs(r); //forward
+    push_channel_attrs(r); //return
+    push_be32(r->b, NFS_PROGRAM);
+    push_be32(r->b, 0); // auth params null
+}
+
+status parse_create_session(client c, buffer b)
+{
+    // check length
+    memcpy(c->session, b->contents + b->start, sizeof(c->session));
+    b->start +=sizeof(c->session);
+    c->sequence = read_beu32(c, b);
+    return STATUS_OK;
+}
+
+void push_sequence(rpc r)
+{
+    push_op(r, OP_SEQUENCE);
+    push_session_id(r, r->c->session);
+    push_be32(r->b, r->c->sequence); 
+    push_be32(r->b, 0x00000000);  // slotid
+    push_be32(r->b, 0x00000000);  // highest slotid
+    push_be32(r->b, 0x00000000);  // sa_cachethis
+    r->c->sequence++;
+}
+
+void push_owner(rpc r)
+{
+    char own[12];
+    push_client_id(r); // 5661 18.16.3 says server must ignore clientid
+    sprintf(own, "open id:%lx", *(unsigned long *)r->c->instance_verifier);
+    push_string(r->b, own, 12);
+}
+
+void push_claim_null(buffer b, buffer name)
+{
+    push_be32(b, CLAIM_NULL);
+    push_string(b, name->contents + name->start, length(name));
+}
+
+void push_claim_deleg_cur(buffer b, buffer name)
+{
+    push_be32(b, CLAIM_DELEGATE_CUR);
+    // state id 4 
+}
+
+void push_open(rpc r, buffer name, boolean create)
+{
+    push_op(r, OP_OPEN);
+    push_be32(r->b, 0); // seqid
+    push_be32(r->b, 2); // share access
+    push_be32(r->b, 0); // share deny
+    push_owner(r);
+    if (create) {
+        push_be32(r->b, OPEN4_CREATE);
+        push_be32(r->b, UNCHECKED4);
+        // xxx - encode this properly, this is the fattr for
+        // the newly created file
+        push_be32(r->b, 0x00000002);
+        push_be32(r->b, 0x00000000);
+        push_be32(r->b, 0x00000002);
+        push_be32(r->b, 0x00000004);
+        push_be32(r->b, 0x000001a4);
+    } else {
+        push_be32(r->b, OPEN4_NOCREATE);
+    }
+    push_claim_null(r->b, name);
+}
+
+void push_stateid(rpc r)
+{
+    // where do we get one of these?
+    push_be32(r->b, 0); // seq
+    push_be32(r->b, 0); // opaque
+    push_be32(r->b, 0); 
+    push_be32(r->b, 0);
+}
+
+// section 18.35, page 494, rfc 5661.txt
+void push_exchange_id(rpc r)
+{
+    // this is supposed to be an instance descriptor that
+    // spans multiple invocations of the same logical instance...
+    // either expose in create_client or assume we will never
+    // try to reclaim locks
+    char co_owner_id[] = "sqlite.ip-172-31-27-113";
```

Comment:
For now I think it's safe to assume that we will never try to reclaim locks. We can treat that as a future feature.